### DISCUSSION TITLE
How does autocomplete output colored text?

### DISCUSSION THREAD

**Discussion by not-poma** [marlonrichert/zsh-autocomplete#555] [General]
I'm trying to build another (unrelated) tool, and want to display colored code under the prompt with `bat` but `zle -R` cannot output newlines or control sequences, so I'm trying to find out how to do it. I see that zsh-autocomplete does it but after some time digging into the code still can't find what part of the code does that. I'm pretty new to zsh widgets and all this terminal manipulation looks pretty arcane to me, at least at first.

**Comment by marlonrichert** [comment]
Autocomplete doesn't do this directly. It simply tells Zsh's completion system what formatting to use for this. Most of it is done [over here](https://github.com/marlonrichert/zsh-autocomplete/blob/f79acc771eeb7ab9e61d731ad210a06b164cf992/scripts/.autocomplete.config#L115).

However, you can find a much simpler, more practical example in my [Zsh Edit plugin](https://github.com/marlonrichert/zsh-edit). Have a look there at the functions `.edit.visual-yank` and `_edit._show_clipboard`, and how they are used in the code.
